lib/Echidna/Exec.hs

```diff
@@ -41,15 +43,45 @@ import EVM.ABI      (AbiValue(..), abiCalldata, abiValueType, encodeAbiValue)
 import EVM.Concrete (Blob(..))
 import EVM.Exec     (exec)
 
-import Echidna.ABI (SolCall, SolSignature, displayAbiCall, encodeSig, genInteractions)
-import Echidna.Internal.Runner
+import Echidna.ABI (SolCall, SolSignature, displayAbiCall, encodeSig, genInteractions, mutateCall)
+import Echidna.Internal.Runner 
 
-type Coverage = MultiSet Int
-type CoverageRef = IORef Coverage
+
+--------------------------------------------------------------------
+-- COVERAGE HANDLING
+
+type CoverageInfo = (SolCall, Set Int)
+type CoverageRef  = IORef CoverageInfo
+
+emptyCovInfo :: CoverageInfo
+emptyCovInfo = (("",[]),empty)
+
+
+getCover :: [CoverageInfo] -> IO [SolCall]
+getCover [] = return []
+getCover xs = setCover vs empty totalCoverage []
+  where vs = V.fromList xs
+        totalCoverage = size $ foldl' (\acc (_,c) -> union acc c) empty xs
+
+setCover :: V.Vector CoverageInfo -> Set Int -> Int -> [SolCall] -> IO [SolCall]
+setCover vs cov tot calls = do
+    let i = maxIndexBy (\a b -> if size (union cov (snd a)) < size (union cov (snd b)) then LT else GT) vs
+        s = vs V.! i
+        c = union cov $ snd s
+        newCalls = (fst s):calls
+
+    if size c == tot
+      then return newCalls
+      else do
+      vs' <- V.unsafeThaw vs
+      M.write vs' i emptyCovInfo
+      res <- V.unsafeFreeze vs'
+      setCover res c tot newCalls
+  
 
 execCallUsing :: MonadState VM m => m VMResult -> SolCall -> m VMResult
 execCallUsing m (t,vs) = cleanUp >> (state . calldata .= cd >> m) where
-  cd = B . abiCalldata (encodeSig t $ abiValueType <$> vs) $ fromList vs
+  cd = B . abiCalldata (encodeSig t $ abiValueType <$> vs) $ V.fromList vs
```

Comment:
Instead of going to/from lists, can we just change everything to use Vectors? I don't think we ever actually want linked lists for reasons other than they're easy. Alternatively, stuff over arbitrary Foldables is even better if that's viable for some functions

Reply:
Update: we push/pop a lot, so we may actually want linked lists instead of vectors here!

Reply:
Here we actually can't use a linked list because `abiCallData` has type sig `Text -> Vector AbiValue -> ByteString`

Reply:
Sorry, the above was unclear. I meant just that my original suggestion was likely unnecessary